tests/jni_api.rs

```diff
@@ -316,6 +319,194 @@ pub fn java_byte_array_from_slice() {
+    env.get_byte_array_region(java_array, 0, &mut res).unwrap();
+    assert_eq!(res[0], 2);
+    assert_eq!(res[1], 3);
+    assert_eq!(res[2], 4);
+}
+
+#[test]
+pub fn get_byte_array_elements_auto() {
+    let env = attach_current_thread();
+
+    // Create original Java array
+    let buf: &[u8] = &[1, 2, 3];
+    let java_array = env
+        .byte_array_from_slice(buf)
+        .expect("JNIEnv#byte_array_from_slice must create a java array from slice");
+
+    // Use a scope to test Drop
+    {
+        // Get byte array elements auto wrapper
+        let auto_ptr = env.get_byte_array_elements_auto(java_array, Copy).unwrap();
+
+        // Check pointer access
+        let ptr = auto_ptr.as_ptr();
+        assert_eq!(unsafe { *ptr.offset(0) }, 1);
+        assert_eq!(unsafe { *ptr.offset(1) }, 2);
+        assert_eq!(unsafe { *ptr.offset(2) }, 3);
+
+        // Check pointer From access
+        let ptr: *mut jbyte = std::convert::From::from(&auto_ptr);
+        assert_eq!(unsafe { *ptr.offset(0) }, 1);
+        assert_eq!(unsafe { *ptr.offset(1) }, 2);
+        assert_eq!(unsafe { *ptr.offset(2) }, 3);
+
+        // Check pointer into() access
+        let ptr: *mut jbyte = (&auto_ptr).into();
+        assert_eq!(unsafe { *ptr.offset(0) }, 1);
+        assert_eq!(unsafe { *ptr.offset(1) }, 2);
+        assert_eq!(unsafe { *ptr.offset(2) }, 3);
+
+        // Modify
+        unsafe {
+            *ptr.offset(0) += 1;
+            *ptr.offset(1) += 1;
+            *ptr.offset(2) += 1;
+        }
+
+        // Commit would be necessary here, if there were no closure
+        //auto_ptr.commit();
+    }
+
+    // Confirm modification of original Java array
+    let mut res: [i8; 3] = [0; 3];
+    env.get_byte_array_region(java_array, 0, &mut res).unwrap();
+    assert_eq!(res[0], 2);
+    assert_eq!(res[1], 3);
+    assert_eq!(res[2], 4);
+}
+
+#[test]
+pub fn java_get_primitive_array_critical() {
+    let env = attach_current_thread();
+
+    // Create original Java array
+    let buf: &[u8] = &[1, 2, 3];
+    let java_array = env
+        .byte_array_from_slice(buf)
+        .expect("JNIEnv#byte_array_from_slice must create a java array from slice");
+
+    // Get array elements
+    let mut is_copy: jboolean = 0xff;
+    let ptr = env
+        .get_primitive_array_critical(java_array, &mut is_copy)
+        .unwrap();
+
+    // Convert void pointer to an unsigned byte array, without copy
+    let mut vec;
+    unsafe { vec = Vec::from_raw_parts(ptr as *mut u8, 3, 3) }
+
+    // Check
+    assert!(is_copy == sys::JNI_FALSE || is_copy == sys::JNI_TRUE);
+    assert_eq!(vec[0], 1);
+    assert_eq!(vec[1], 2);
+    assert_eq!(vec[2], 3);
+
+    // Modify
+    vec[0] += 1;
+    vec[1] += 1;
+    vec[2] += 1;
+
+    // Release
+    // First, make sure vec's destructor doesn't free the data it thinks it owns when it goes out
+    // of scope (so release_primitive_array_critical() can properly free it)
+    std::mem::ManuallyDrop::new(vec);
```

Comment:
I think that it is more idiomatic to use `ManuallyDrop` when creating a vector. Otherwise it is easier to use `mem::forget` here.

Reply:
OK. Thanks.